### PR TITLE
define cache for seed file in seed pattern

### PR DESCRIPTION
**Pull request by wangforthinker**
Signed-off-by: allen.wq <allen.wq@alipay.com>

### Ⅰ. Describe what this PR did
Add implementation for seed cache. Please see refer #1284.

### Ⅱ. Does this pull request fix one issue?

NONE.

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
Add unit test.

### CODE REVIEW

**Comment by hhhhsdxxxx** [comment]
lgtm!
Reviewed-by henry.hj@antfin.com

**Comment by xujihui1985** [comment]
lgtm
Reviewed-by jihui.xjh@antfin.com

File: dfdaemon/seed/cache.go [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+// syncMemoryCache flushes the memory cache to local file.
+func (fcb *fileCacheBuffer) syncMemoryCache() error {
+	var (
+		err error
+	)
+
+	offArr := []int64{}
+	bufArr := []*bytes.Buffer{}
+
+	fcb.RLock()
```
[the remark this answers is not in the document]
Comment (wangforthinker) [reply dfdaemon/seed/cache.go:245]: the elem of memCacheMap represents a buffer cache,  if buffer cache stores in  memCacheMap, its content could  not be changed util buffer gc. 
So the Lock aims to protect the memCacheMap not the buffer content.

File: dfdaemon/seed/cache.go [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+
+	defer func() {
+		if err != nil {
+			fw.Close()
+		}
+	}()
+
+	fcb := &fileCacheBuffer{path: path, fw: fw, fullSize: fullSize, memoryCache: memoryCache}
+	if memoryCache {
+		fcb.blockOrder = blockOrder
```
[the remark this answers is not in the document]
Comment (wangforthinker) [reply dfdaemon/seed/cache.go:85]: Add check for it .

File: dfdaemon/seed/cache.go [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+	bufArr := []*bytes.Buffer{}
+
+	fcb.RLock()
+	for off, buf := range fcb.memCacheMap {
+		offArr = append(offArr, off)
+		bufArr = append(bufArr, buf)
+	}
+	fcb.RUnlock()
+
+	for i := 0; i < len(offArr); i++ {
```
[the remark this answers is not in the document]
Comment (wangforthinker) [reply dfdaemon/seed/cache.go:256]: done.

File: dfdaemon/seed/cache.go [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+		if (fullSize % int64(fcb.blockSize)) > 0 {
+			blocks++
+		}
+
+		sizeOf64Bits := uint32(blocks / 64)
+		if (blocks % 64) > 0 {
+			sizeOf64Bits++
+		}
+
+		fcb.blockMeta, err = bitmap.NewBitMap(sizeOf64Bits, false)
```
[the remark this answers is not in the document]
Comment (wangforthinker) [reply dfdaemon/seed/cache.go]: done.

File: dfdaemon/seed/cache.go [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+
+	if fcb.remove {
+		return 0, 0, io.ErrClosedPipe
+	}
+
+	if off < 0 {
+		off = 0
+	}
+
+	// if size <= 0, set range to [off, fullSize-1]
```
[the remark this answers is not in the document]
Comment (lowzj) [reply dfdaemon/seed/cache.go:204]: So it means that the empty data is not supported here.

File: dfdaemon/seed/cache.go [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+		bufArr = append(bufArr, buf)
+	}
+	fcb.RUnlock()
+
+	sort.Slice(offArr, func(i, j int) bool {
+		return offArr[i] < offArr[j]
+	})
+
+	for i := 0; i < len(offArr); i++ {
+		err = fcb.syncBlockToFile(bufArr[i].Bytes(), offArr[i])
```
[the remark this answers is not in the document]
Comment (wangforthinker) [reply dfdaemon/seed/cache.go]: Fix it by add <off, buf> to arrary.

**Pull request merged**